tests/integration/servicemesh/util.go

```diff
@@ -0,0 +1,123 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servicemesh
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	// import maistra CRD manifests
+	_ "maistra.io/api/manifests"
+
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/cluster"
+	"istio.io/istio/pkg/test/framework/resource"
+	kubetest "istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+var (
+	manifestsDir      = env.IstioSrc + "/vendor/maistra.io/api/manifests"
+	bookinfoManifests = []string{
+		env.IstioSrc + "/samples/bookinfo/platform/kube/bookinfo.yaml",
+		env.IstioSrc + "/samples/bookinfo/networking/bookinfo-gateway.yaml",
+	}
+	sleepManifest = env.IstioSrc + "/samples/sleep/sleep.yaml"
+	rnd           = rand.New(rand.NewSource(time.Now().UnixNano()))
+)
+
+func ApplyServiceMeshCRDs(ctx resource.Context) (err error) {
+	crds, err := findCRDs()
+	if err != nil {
+		return fmt.Errorf("cannot read maistra CRD YAMLs: %s", err)
+	}
+	for _, cluster := range ctx.Clusters().Kube().Primaries() {
+		if err = cluster.ApplyYAMLFiles("", crds...); err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func findCRDs() (list []string, err error) {
+	list = []string{}
+	files, err := ioutil.ReadDir(manifestsDir)
+	if err != nil {
+		return
+	}
+	for _, file := range files {
+		if !file.IsDir() && strings.HasSuffix(file.Name(), ".yaml") {
+			list = append(list, file.Name())
+		}
+	}
+	return
+}
+
+func InstallBookinfo(ctx framework.TestContext, c cluster.Cluster, namespace string) {
+	if err := c.ApplyYAMLFiles(namespace, bookinfoManifests...); err != nil {
+		ctx.Fatal(err)
+	}
+	if err := retry.UntilSuccess(func() error {
+		if _, err := kubetest.CheckPodsAreReady(kubetest.NewPodFetch(c, namespace, "app=ratings")); err != nil {
+			return fmt.Errorf("ratings pod is not ready: %v", err)
+		}
+		return nil
+	}, retry.Timeout(300*time.Second), retry.Delay(time.Second)); err != nil {
+		ctx.Fatal(err)
+	}
+}
+
+func InstallSleep(ctx framework.TestContext, c cluster.Cluster, namespace string) {
+	if err := c.ApplyYAMLFiles(namespace, sleepManifest); err != nil {
+		ctx.Fatal(err)
+	}
+	if err := retry.UntilSuccess(func() error {
+		if _, err := kubetest.CheckPodsAreReady(kubetest.NewPodFetch(c, namespace, "app=sleep")); err != nil {
+			return fmt.Errorf("sleep pod is not ready: %v", err)
+		}
+		return nil
+	}, retry.Timeout(300*time.Second), retry.Delay(time.Second)); err != nil {
+		ctx.Fatal(err)
+	}
+}
+
+// TODO for some reason namespace.NewOrFail() doesn't work so I'm doing this manually
```

Comment:
weird, no clue at all? no logs? worth to create a ticket and add it to OSSM-401?

Reply:
it always threw `namespace already exists`. no idea why

Reply:
`namespace.NewOrFail()` seems to loop through clusters (see [1]) and creates the namespace in each of them. Maybe the same cluster appears twice in the clusters list. If so, this could also be causing other issues. 

[1] https://github.com/Maistra/istio/blob/c45e902630dcb10138533a3ee255dd844c290ab0/pkg/test/framework/components/namespace/kube.go#L159

Reply:
that's right, but I'm not setting up the list of clusters, that is coming from upstream